game/scripts/vscripts/abilities/sohei/sohei_guard.lua

```diff
@@ -0,0 +1,370 @@
+sohei_guard = class( AbilityBaseClass )
+
+LinkLuaModifier( "modifier_sohei_guard_reflect", "abilities/sohei/sohei_guard.lua", LUA_MODIFIER_MOTION_NONE )
+LinkLuaModifier( "modifier_sohei_guard_knockback", "abilities/sohei/sohei_guard.lua", LUA_MODIFIER_MOTION_HORIZONTAL )
+
+--------------------------------------------------------------------------------
+
+-- unfinished talent stuff
+function sohei_guard:CastFilterResultTarget( target )
+	local caster = self:GetCaster()
+
+	if ( target ~= caster ) and caster:IsStunned() then
+		return UF_FAIL_CUSTOM
+	end
+
+	local ufResult = UnitFilter(
+		target,
+		self:GetAbilityTargetTeam(),
+		self:GetAbilityTargetType(),
+		self:GetAbilityTargetFlags(),
+		caster:GetTeamNumber()
+	)
+
+	return ufResult
+end
+
+--------------------------------------------------------------------------------
+
+-- unfinished talent stuff
+function sohei_guard:GetCustomCastErrorTarget( target )
+	local caster = self:GetCaster()
+
+	if ( target ~= caster ) and caster:IsStunned() then
+		return "#dota_hud_error_cant_cast_on_ally_while_stunned"
+	end
+
+	return ""
+end
+
+--------------------------------------------------------------------------------
+
+if IsServer() then
+	-- always preferrable to stop a cast instead of faking not casting
+	function sohei_guard:GetBehavior()
+		local behavior = self.BaseClass.GetBehavior( self )
+		local caster = self:GetCaster()
+		local modifier_charges = caster:FindModifierByName( "modifier_sohei_dash_charges" )
+		local talent = caster:FindAbilityByName( "special_bonus_sohei_guard_allycast" )
+
+		-- unfinished talent stuff
+		if talent and talent:GetLevel() > 0 then
+			behavior = bit.bor( DOTA_ABILITY_BEHAVIOR_UNIT_TARGET, DOTA_ABILITY_BEHAVIOR_IMMEDIATE, DOTA_ABILITY_BEHAVIOR_DONT_CANCEL_CHANNEL )
+		end
+
+		if modifier_charges and modifier_charges:GetStackCount() >= 2 then
+			behavior = bit.bor( behavior, DOTA_ABILITY_BEHAVIOR_IGNORE_PSEUDO_QUEUE )
+		end
+
+		return behavior
+	end
+
+--------------------------------------------------------------------------------
+
+	function sohei_guard:OnSpellStart()
+		local caster = self:GetCaster()
+		local target = self:GetCursorTarget() or caster
+
+		-- Check if there are enough charges to cast the ability, if the caster is stunned
+
+		if caster:IsStunned() then
+			local modifier_charges = caster:FindModifierByName( "modifier_sohei_dash_charges" )
+
+			if modifier_charges then
+				modifier_charges:SetStackCount( modifier_charges:GetStackCount() - 2 )
```

Comment:
Like I said in the comments ( scattered though they may be ), OnSpellStart is really much too late to be doing these types checks. Too much stuff has already happened at this point in the logic chain to say no to the ability happening, so imo it's really better to just let the ability fire off if the spell somehow reached this point even if the conditions aren't right to avoid stuff like what happened with the previous Dash talent bug.

Reply:
Oh i see, i didn't notice the code to put the two abilities on cooldown.

Do you think implementing RefreshCharges on the ability will make refresher work with it? I know it'll make it work with duels and our custom refreshers at least, we call that method...

Reply:
Looks like it doesn't work with vanilla orb but works with everything else, though the code I set up didn't accommodate for other sources of charge gaining so I'm redoing that part of the code a bit too.